k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/ImapFolder.java

```diff
@@ -1143,64 +1144,164 @@ private void parseBodyStructure(ImapList bs, Part part, String id) throws Messag
      */
     @Override
     public Map<String, String> appendMessages(List<? extends Message> messages) throws MessagingException {
+        if (messages == null || messages.size() == 0) {
+            return null;
+        }
+
         open(OPEN_MODE_RW);
         checkOpen();
 
         try {
             Map<String, String> uidMap = new HashMap<>();
-            for (Message message : messages) {
-                long messageSize = message.calculateSize();
 
-                String encodeFolderName = folderNameCodec.encode(getPrefixedName());
-                String escapedFolderName = ImapUtility.encodeString(encodeFolderName);
-                String command = String.format(Locale.US, "APPEND %s (%s) {%d}", escapedFolderName,
-                        combineFlags(message.getFlags()), messageSize);
-                connection.sendCommand(command, false);
+            if (connection.isMultiappendCapable() && messages.size() > 1) {
+                appendWithMultiappend(messages, uidMap);
+            } else {
+                appendWithoutMultiappend(messages, uidMap);
+            }
 
-                ImapResponse response;
-                do {
-                    response = connection.readResponse();
+            /*
+             * We need uidMap to be null if new UIDs are not available to maintain consistency
+             * with the behavior of other similar methods (copyMessages, moveMessages) which
+             * return null.
```

Comment:
We should really change this. Returning null is bad and an empty map would suffice.

Reply:
I'd leave this for another PR, since this touches other methods besides append